### ISSUE TITLE
armhf+arm64: locale-gen aborts when running via qemu on artful and bionic

### ISSUE THREAD

**Issue by nuclearsandwich** [osrf/multiarch-docker-image-generation#17]
This is a reference to the upstream issues reported in qemu and glibc.

[qemu issue](https://bugs.launchpad.net/qemu/+bug/1673976)
[glibc issue](https://sourceware.org/bugzilla/show_bug.cgi?id=22273)

It can be reproduced with a quick

```
$ docker run -ti osrf/ubuntu_arm64:artful bash
# echo 'en_US.UTF-8 UTF-8' >> /etc/locale.gen
# locale-gen
```

And is causing builds to fail for Melodic [on armhf](http://build.ros.org/view/Mbin_ubhf_uBhf/job/Mbin_ubhf_uBhf__catkin__ubuntu_bionic_armhf__binary/4/console#console-section-9) and [arm64](http://build.ros.org/view/Mbin_ubv8_uBv8/job/Mbin_ubv8_uBv8__catkin__ubuntu_bionic_arm64__binary/2/console#console-section-9)

Since there was an unsubstantiated comment that the issue might have been introduced in qemu 2.8 I tried this out with an artful image and the old qemu 2.1 binary and got the same assert. So rolling back #14 for artful and bionic won't work as a tempfix.

**Comment by clalancette** [comment]
This is the commit that theoretically fixes it in glibc: https://sourceware.org/git/?p=glibc.git;a=commit;h=fe05e1cb6d64dba6172249c79526f1e9af8f2bfd

**Comment by nuclearsandwich** [comment]
I have an image that tries the minimal workaround of gunzipping `/usr/share/i18n/charmaps/UTF-8.gz` before running locale-gen and am running a local build to see if that's sufficient for a catkin binarydeb.

**Comment by nuclearsandwich** [comment]
So I was able to build catkin for arm64 locally by replacing osrf/ubuntu_arm64:bionic with an image that has an uncompressed UTF-8.

If we can't get anything else going in the short term, we could deploy that change to our images for a very narrow fix.

Currently trying to patch glibc 2.26 from bionic sourcedeb.

**Comment by tfoote** [comment]
It looks like a 2.27 release is due Feb 1st, but that likely won't be backported so a patch against the deb sources of that specific patch might be the best approach against 2.26-4 in the debian tree: https://anonscm.debian.org/cgit/pkg-glibc/glibc.git/log/?h=debian/2.26-4

**Comment by clalancette** [comment]
I ended up doing something like what @tfoote suggested, though I based it off of the 2.26-0ubuntu3 package that Ubuntu bionic is currently shipping.  In particular, I backported https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=fe05e1cb6d64dba6172249c79526f1e9af8f2bfd and https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=aa95a2414e4f664ca740ad5f4a72d9145abbd426 , and rebuilt an ARM64 package with that.  In testing, this seems to fix the problem for now.  We are going to do some additional testing, and I'm going to go ahead and build an armhf version of this package for more testing, but it looks like we are on the right track.

**Comment by nuclearsandwich** [comment]
I injected @clalancette's debs into a local release script run building ros-melodic-catkin for bionic arm64 and it was successful.

Since our Packet.net host doesn't appear to be able to run native armhf instructions, @clalancette is building armhf on a local device.

The next steps are in parallel

* Deploy this to unblock arm builds on our farm.
* Propose these changes for inclusion in upstream Debian and Ubuntu so we do not need to maintain them. There is some urgency to this as the Debian import freeze for bionic is March 1. If we want to get both Debian and Ubuntu shipping the same version of the changes we need to get it accepted in both before then.

For deployment, the simplest thing to do would be to commit my patched containers and push them as the new `osrf/ubuntu_arm{64,hf}:bionic`. The buildfarm and any downstream users would pick up the changes transparently. But if there are any regressions or other issues, they will hit everyone at once.

**Comment by tfoote** [comment]
That seems like a reasonable approach. Committing the patched ones and pushing for a test deployment makes sense, keep a backup of the old image locally and it can be quickly rolled back. But we should make sure to integrate applying that patch into this repo so when we next regenerate the patched version will be rolled in without a manual process.

**Comment by clalancette** [comment]
Just to give all of the details of how I built these packages:

* Get access to a native ARM64 machine.  The bigger the better, since the build takes an hour even in the best of cases.  Also, make sure the ARM64 machine supports 32-bit ARMHF, as not all of them do.
* Create a chroot for each arch in `arm64, armhf`:
  * `mkdir -p /var/lib/chroot/bionic_$arch`
  * `debootstrap --variant=buildd --arch=$arch bionic /var/lib/chroot/bionic_$arch http://ports.ubuntu.com/ubuntu-ports`
  * `chroot /var/lib/chroot/bionic_$arch`
  * `mount -t proc none /proc`
  * `mount -t sysfs none /sys`
  * `echo 'deb http://ports.ubuntu.com/ubuntu-ports bionic main restricted universe multiverse' > /etc/apt/sources.list`
  * `echo 'deb-src http://ports.ubuntu.com/ubuntu-ports bionic main restricted universe multiverse' >> /etc/apt/sources.list`
  * `apt-get update`
  * `apt-get -y install devscripts vim quilt autoconf gawk debhelper rdfind symlinks gperf systemtap-sdt-dev libaudit-dev libcap-dev libselinux-dev g++-6 po-debconf`
  * `exit`
* Download the upstream Ubuntu 2.26-0ubuntu3 tarballs:
  * `wget https://launchpad.net/ubuntu/+archive/primary/+files/glibc_2.26.orig.tar.xz`
  * `wget https://launchpad.net/ubuntu/+archive/primary/+files/glibc_2.26-0ubuntu3.debian.tar.xz`
* Copy the tarballs into the chroots:
  * `cp glibc_2.26.orig.tar.xz glibc_2.26-0ubuntu3.debian.tar.xz /var/lib/chroot/bionic_arm64/root`
  * `cp glibc_2.26.orig.tar.xz glibc_2.26-0ubuntu3.debian.tar.xz /var/lib/chroot/bionic_armhf/root`
* At this point, it is possible to compile the packages (instructions on how to do this below).  However, note that even with these unmodified, upstream versions of the packages, I had test failures near the end of the build.  In particular, there are a set of 4 `tst-cancel` tests (`tst-cancel4`, `tst-cancel5`, `tst-cancelx4`, `tst-cancelx5`) that always seem to fail with "send returned" errors.  Because these were in the unmodified code, I decided to just disable these tests.  We could look into this further, but it is probably something environmental.
* Next, apply the fixes from upstream.  I applied the patches from upstream glibc linked above; the details of how to apply to a debian package are left to the reader.  Also don't forget to disable the `tst-cancel` tests by modifying `debian/testsuite-xfail-debian.mk`.
* Now build the packages.  For each arch in `arm64, armhf`, do:
  * `chroot /var/lib/chroot/bionic_$arch`
  * `cd`
  * `tar -xvf glibc_2.26.orig.tar.xz`
  * `cd glibc-2.26`
  * `tar -xvf ../glibc_2.26-0ubuntu3osrf1.debian.tar.xz`
  * `debuild -uc -us`

**Comment by nuclearsandwich** [comment]
It's not pretty or clean, but a working osrf/ubuntu_arm64:bionic has been pushed. There's a copy of the clean image from last week is pushed to Docker Hub as `nuclearsandwich/ubuntu_arm64:bionic_clean`.

I'm going to open a PR that sticks @clalancette's debs in a branch but I'm going to focus on getting the changes upstreamed before working on integrating the debs into our image build pipeline.

**Comment by nuclearsandwich** [comment]
Deployed the armhf docker image as well and builds are running for armhf bionic. Again the base armhf docker image is still available at `nuclearsandwich/ubuntu_armhf:bionic`.

I'll next be working on proposing these changes upstream.

**Comment by nuclearsandwich** [comment]
Good news! The upstream debian package actually already includes these patches in the 2.26-6 release from Dec 31 and I just [couldn't find the patch](https://anonscm.debian.org/cgit/pkg-glibc/glibc.git/commit/?h=d52a330c) in the giant diff file.

Incidentally it's in that giant diff because it has also been backported upstream by glibc into 2.26: [initial assert fix](https://sourceware.org/git/?p=glibc.git;a=commit;h=f8ee700e8959236bb2c54f3aacf57edca5dab186) and [wnohang followup](https://sourceware.org/git/?p=glibc.git;a=commit;h=bfdb34f2f26933a831684c09cadc2b782f7fbd7a).

So we just need to ask the Ubuntu folks to sync it down.

**Comment by nuclearsandwich** [comment]
through no action on our part, the patches are now part of [glibc2.26-0ubuntu4](https://launchpad.net/ubuntu/+source/glibc/2.26-0ubuntu4) (verified by checking changelog and package sources, the patches are applied in the patch file `git-updates.diff`.

This package is currently in bionic-proposed. I've verified locally that it resolves our issue. I don't think it's currently worth re-rolling images with glibc from bionic-proposed. We should be able to wait for it to filter down and rebuild our images.

**Comment by clalancette** [comment]
Agreed.  Now that the ticking time bomb has been defused, I think it is safe to stay with our packages for now, and then upgrade later when we have another reason to do so.

Given that, should we close this out?

**Comment by nuclearsandwich** [comment]
> Given that, should we close this out?

Because the images aren't currently automatically reproducible, I think it's worth leaving it open until we can rebuild from bionic sources or add the installation of our custom debs to the image baking process.

**Comment by nuclearsandwich** [comment]
A version of glibc 2.27 has landed in bionic and fixes this issue. I am afk and haven't started building new bionic images but when I do I'll close out this issue.

**Comment by nuclearsandwich** [comment]
It doesn't appear that a backported glibc was released for artful. This didn't matter too much to us because we aren't building arm packages for artful. This particular issue is resolved in bionic but there is a new one (#18) blocking image building. So while I haven't built this yet the issue is closed. I also propose dropping artful arm images in https://github.com/osrf/multiarch-docker-image-generation/pull/20
